worker/uniter/relation/state.go

```diff
@@ -108,200 +107,196 @@ func (s *State) Validate(hi hook.Info) (err error) {
 	isApp := false
 	if hi.RemoteUnit == "" {
 		isApp = true
-		name = hi.RemoteApplication + "-app"
 	}
-	path := filepath.Join(d.path, name)
+	// Get a copy of current state to modify, so we only update current
+	// state if the new state was written successfully.
+	newState := s.state.copy()
 	if hi.Kind == hooks.RelationDeparted {
-		if err = os.Remove(path); err != nil && !os.IsNotExist(err) {
-			return err
-		}
-		// If atomic delete succeeded, update own state.
+		// Update own state.
 		if isApp {
-			delete(d.state.ApplicationMembers, hi.RemoteApplication)
+			delete(newState.ApplicationMembers, hi.RemoteApplication)
 		} else {
-			delete(d.state.Members, hi.RemoteUnit)
+			delete(newState.Members, hi.RemoteUnit)
 		}
-		return nil
+		return s.writeAndUpdateLocal(*newState)
 	}
-	di := diskInfo{&hi.ChangeVersion, hi.Kind == hooks.RelationJoined}
-	if err := utils.WriteYaml(path, &di); err != nil {
-		return err
-	}
-	// If write was successful, update own state.
+	// Update own state.
 	if isApp {
-		d.state.ApplicationMembers[hi.RemoteApplication] = hi.ChangeVersion
+		newState.ApplicationMembers[hi.RemoteApplication] = hi.ChangeVersion
 	} else {
-		d.state.Members[hi.RemoteUnit] = hi.ChangeVersion
+		newState.Members[hi.RemoteUnit] = hi.ChangeVersion
 	}
 	if hi.Kind == hooks.RelationJoined {
-		d.state.ChangedPending = hi.RemoteUnit
+		newState.ChangedPending = hi.RemoteUnit
 	} else {
-		d.state.ChangedPending = ""
+		newState.ChangedPending = ""
+	}
+	return s.writeAndUpdateLocal(*newState)
+}
+
+// writeAndUpdateLocal, writes the new State, if success, the current State is updated.
+func (s *StateOps) writeAndUpdateLocal(newState State) error {
+	if s.state.RelationId != newState.RelationId {
+		return fmt.Errorf("expected relation %d, got relation %d", s.state.RelationId, newState.RelationId)
+	}
+	data, err := yaml.Marshal(newState)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	rs := map[int]string{s.state.RelationId: string(data)}
+	err = s.unitStateRW.SetState(params.SetUnitStateArg{RelationState: &rs})
+	if err != nil {
+		return errors.Trace(err)
 	}
+	s.state = newState
 	return nil
 }
 
-// Remove removes the directory if it exists and is empty.
-func (d *StateDir) Remove() error {
+// Remove removes the persisted state if it inState and is empty.
+func (s *StateOps) Remove() error {
 	// Note(jam): 2019-10-22 os.Remove() requires the directory to be empty, but
 	//  we added "foo-app" but we won't call RelationDeparted for "foo" and thus won't
 	//  delete "foo-app". Instead, during relation-broken, we cleanup all related applications.
-	for appMember := range d.state.ApplicationMembers {
-		path := filepath.Join(d.path, appMember+"-app")
-		if err := os.Remove(path); err != nil && !os.IsNotExist(err) {
-			return errors.Trace(err)
-		}
+	//
+	// Note(hml): 2020-04-01 We no longer keep persisted state as local files,
+	// instead remove all ApplicationMembers.
+	s.state.ApplicationMembers = nil
```

Comment:
I believe this should be at the bottom of the method after L269; we don't want to reset this map before the `SetState` call returns successfully